File: src/1161.最大层内元素和.py
#
# @lc app=leetcode.cn id=1161 lang=python3
#
# [1161] 最大层内元素和
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from queue import Queue


class Solution:
    def maxLevelSum(self, root: TreeNode) -> int:
        q = Queue()
        q.put(root)
        level = 1
        max_level_sum = float('-inf')
        while q.qsize():
            n = q.qsize()
            level_sum = 0
            for _ in range(n):
                node = q.get()
                level_sum += node.val
                if node.left:
                    q.put(node.left)
                if node.right:
                    q.put(node.right)
            if level_sum > max_level_sum:
                max_level_sum = level_sum
                max_level = level
            level += 1
        return max_level

# @lc code=end

File: src/1161.最大层内元素和.rs
/*
 * @lc app=leetcode.cn id=1161 lang=rust
 *
 * [1161] 最大层内元素和
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}
impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
struct Solution;
// @lc code=start
//
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;
impl Solution {
    pub fn max_level_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let (mut max_level, mut max_level_sum) = (1, i32::min_value());
        let mut level = 1;
        let mut q = VecDeque::new();
        q.push_back(root.unwrap());
        //
        while q.len() > 0 {
            let n = q.len();
            let mut level_sum = 0;
            for _ in 0..n {
                let node = q.pop_front().unwrap();
                level_sum += node.borrow().val;
                if let Some(left) = &node.borrow().left {
                    q.push_back(left.clone());
                }
                if let Some(right) = &node.borrow().right {
                    q.push_back(right.clone());
                };
            }
            if level_sum > max_level_sum {
                max_level_sum = level_sum;
                max_level = level;
            }
            level += 1;
        }

        max_level
    }
}
// @lc code=end
